src/celeritas/optical/ScintillationGenerator.hh

```diff
@@ -0,0 +1,206 @@
+/*!
+ * Construct from shared scintillation data and distribution parameters.
+ */
+CELER_FUNCTION
+ScintillationGenerator::ScintillationGenerator(
+    OpticalDistributionData const& dist,
+    NativeCRef<ScintillationData> const& shared,
+    Span<OpticalPrimary> photons)
+    : dist_(dist)
+    , spectrum_(shared.spectrum[dist_.material])
+    , photons_(photons)
+    , sample_cost_(-1, 1)
+    , sample_phi_(0, 2 * constants::pi)
+    , sample_time_(real_type{1} / spectrum_.fall_time[0])
+    , is_neutral_{dist_.charge == zero_quantity()}
+{
+    CELER_EXPECT(dist_);
+    CELER_EXPECT(photons_.size() == dist_.num_photons);
+
+    auto const& pre_step = dist_.points[StepPoint::pre];
+    auto const& post_step = dist_.points[StepPoint::post];
+    delta_pos_ = post_step.pos - pre_step.pos;
+    delta_speed_ = post_step.speed - pre_step.speed;
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Sample scintillation photons from optical property data and step data.
+ *
+ * The optical photons are generated evenly along the step and are emitted
+ * uniformly over the entire solid angle with a random linear polarization.
+ * The photon energy is calculated by the scintillation emission wavelength
+ * \f[
+   E = \frac{hc}{\lambda},
+ * \f]
+ * where \f$ h \f$ is the Planck constant and \f$ c \f$ is the speed of light,
+ * and \f$ \lambda \f$ is sampled by the normal distribution with the mean of
+ * scintillation emission spectrum and the standard deviation. The emitted time
+ * is simulated according to empirical shapes of the material-dependent
+ * scintillation time structure with one or double exponentials.
+ */
+template<class Generator>
+CELER_FUNCTION Span<OpticalPrimary>
+ScintillationGenerator::operator()(Generator& rng)
+{
+    // Loop for generating scintillation photons
+    size_type num_generated{0};
+
+    for (auto sid : range(spectrum_.size))
+    {
+        // Skip if there is no emission warelength
+        if(spectrum_.lambda_mean[sid] <= 0) 
+           continue;
+
+        NormalDistribution<real_type> sample_lambda(
+            spectrum_.lambda_mean[sid], spectrum_.lambda_sigma[sid]);
+
+        size_type num_photons = dist_.num_photons * spectrum_.yield_prob[sid];
+
+        CELER_EXPECT(num_generated + num_photons <= dist_.num_photons);
+
+        for (size_type i : range(num_generated, num_generated + num_photons))
+        {
+            // Sample wavelength and convert to energy
+            real_type wave_length = sample_lambda(rng);
+            CELER_EXPECT(wave_length > 0);
+            photons_[i].energy 
+                = native_value_to<Energy>(this->hc() / wave_length);
+
+            // Sample direction
+            real_type cost = sample_cost_(rng);
+            real_type phi = sample_phi_(rng);
+            photons_[i].direction = rotate(from_spherical(cost, phi), 
+                                           dist_.points[StepPoint::post].pos);
```

Comment:
I think since the directions are sampled randomly (uniformly distributed points on the unit sphere) it can just be `photons_[i].direction = from_spherical(cost, phi)`.

Reply:
Right, it should be just a random isotropic and no need a reference direction(i.e., rotational invariant). Fixed.